### ISSUE TITLE
bug: It seems that we can not override some lazyvim default plugins' keymaps?

### ISSUE THREAD

**Issue by afzw** [LazyVim/LazyVim#1485]
### Did you check docs and existing issues?

- [X] I have read all the LazyVim docs
- [X] I have searched the existing issues of LazyVim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

0.9.1

### Operating system/version

MacOS 13.3

### Describe the bug

I want to move my cursor left/right in insert mode, and I custom my keymaps like so:

![image](https://github.com/LazyVim/LazyVim/assets/50651678/2359b57d-46ef-4580-a113-8c9794a00891)

But it just can not worko(╥﹏╥)o. anyone knows the reason??

（ps: It really works well before when I haven't use lazyvim, is the lazyvim problem?）

### Steps To Reproduce

1. open `./lua/config/keymaps.lua`
2. write `map("i", "<C-b>", "<LEFT>", opt)`
3. reload the nvim and the keymap can not work -- I got `^B` when I put `<C-b>`.

### Expected Behavior

The <C-b>, <C-f> keymaps in Insert mode work well.

### Repro

```Lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath, })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
  "folke/tokyonight.nvim",
  "folke/LazyVim",
  -- add any other plugins here
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```

**Comment by dpetka2001** [comment]
With fresh Lazyvim starter installation using default configuration, I put the following in `/lua/config/keymaps.lua`
```lua
vim.keymap.set("i", "<C-b>", "<LEFT>", { desc = "LEFT" })
```
and it worked fine for me as long as you don't have `nvim-cmp` completion menu open, because when that menu is open `<C-b>` is used for scrolling backwards through the docs.

**Comment by afzw** [comment]
> With fresh Lazyvim starter installation using default configuration, I put the following in `/lua/config/keymaps.lua`
> 
> ```lua
> vim.keymap.set("i", "<C-b>", "<LEFT>", { desc = "LEFT" })
> ```
> 
> and it worked fine for me as long as you don't have `nvim-cmp` completion menu open, because when that menu is open `<C-b>` is used for scrolling backwards through the docs.

Thanks for your reply!!
But the problem remains, I just got `^B` when I put `<C-b>`...

**Comment by dpetka2001** [comment]
What does `verbose imap <C-b>` say? There must be some kind of conflict in your configuration, since it doesn't happen with Lazyvim default installation.

**Comment by afzw** [comment]
> What does `verbose imap <C-b>` say? There must be some kind of conflict in your configuration, since it doesn't happen with Lazyvim default installation.

Here the output:

<img width="1026" alt="image" src="https://github.com/LazyVim/LazyVim/assets/50651678/eac00d3f-df80-4086-a7a1-3e117ee65fb1">

oh well, it seems sth. in the path override my keymaps...

I find out! it's the `noice.nvim` keymaps !

now I'm still a little confused：
1. in my opinion, what I change in the `/lua/config/keymaps.lua` has the highest priority which will override all the default keymaps the lazyvim provide, so why is it happen? (is my opinion wrong?)
2. how to fix it elegantly??

**Comment by dpetka2001** [comment]
What is your `map` function bound to? Is it set `local map = vim.keymap.set`?

**Comment by afzw** [comment]
> What is your `map` function bound to? Is it set `local map = vim.keymap.set`?

yes, I set `map` to `vim.keymap.set` and the offical code below:
```lua
local function map(mode, lhs, rhs, opts)
  local keys = require("lazy.core.handler").handlers.keys
  if not keys.active[keys.parse({ lhs, mode = mode }).id] then
    opts = opts or {}
    opts.silent = opts.silent ~= false
    if opts.remap and not vim.g.vscode then
      opts.remap = nil
    end
    vim.keymap.set(mode, lhs, rhs, opts)
  end
end
```
but neither works...

**Comment by dpetka2001** [comment]
Do you have a link on github to your nvim dotfiles?

**Comment by afzw** [comment]
> Do you have a link on github to your nvim dotfiles?

here~~:
https://github.com/afzw/confighub

**Comment by dpetka2001** [comment]
I tried with your nvim dotfiles and it works for me. `<C-b>` goes `left` and `<C-f>` goes right. Do you have some kind of mapping with these keys in your terminal settings that might conflict with Lazyvim's keymaps?

**Comment by afzw** [comment]
> I tried with your nvim dotfiles and it works for me. `<C-b>` goes `left` and `<C-f>` goes right. Do you have some kind of mapping with these keys in your terminal settings that might conflict with Lazyvim's keymaps?

i dont think so,,, well, as I mentioned at the first, I used to do it well without lazyvim.
later I switch to lazyvim and got the problem...😭

**Comment by dpetka2001** [comment]
I would suggest you to try with a terminal like `wezterm` or `kitty` with default configuration and observe how it behaves.

**Comment by afzw** [comment]
> I would suggest you to try with a terminal like `wezterm` or `kitty` with default configuration and observe how it behaves.

I have tried using the default terminal of macos, also didn't work..😂

The problem has to be somewhere else...

**Comment by afzw** [comment]
> I have read several problems with how key codes get sent throught the terminal on MacOS, that's why I suggested those 2. If you want, you could test it out just to remove the possibility that it has something to do with your terminal. Then you know you have to search elsewhere.

Oh I see , ok, I'll try, I'll give the result later after I try those 2 ~~

thank you again!!

**Comment by afzw** [comment]
well, can't work...@dpetka2001

**Comment by afzw** [comment]
Magic ?!
It must be magic !
You guys guess what ?!
I just make it but only once ??!

Well, I think I get it  —— when some UI-about  plugins lazyVim provided loaded, the keymaps will go wrong, otherwise, it goes well. I think it is the `noice.nvim` that make my keymaps breakdown. I'll find out

**Comment by afzw** [comment]
Yes, gotcha !
When I disabled the `noice` plugin, the <C-b>, <C-f> keymaps go well 🎉 !

It looks like that my keymaps didn't override the lazyVim default pulgins' keymaps, is it right??

**Comment by afzw** [comment]
I found out that I can not override the plugins' keymaps which lazyvim provided and **`lazy = true` or `event = "VeryLazy"`**

I don't know much about lazyvim logic, I guess that my custom keymaps were overrided by the "lazy" plugins' keymaps because they would load later than my keymaps.

**Comment by dpetka2001** [comment]
I understand what you're saying and you're right about that, but I tried your dotfiles and worked fine for me without changing anything in your dotfiles. `noice` is installed by default in Lazyvim, so I didn't change anything. It should behave the same for me as for you. The keymaps were working as intended and indeed the keymaps in `/lua/config/keymaps.lua` should take priority I think (not 100% sure).

**Comment by dpetka2001** [comment]
@afzw I think I found out what's wrong. `keymaps` will only load on `VeryLazy` when you open `nvim` without any arguments. If you do `nvim file.lua` they won't load on `VeryLazy`, thus plugins that load on `VeryLazy` event overwrite the keymaps. You can try this by opening Neovim with just `nvim` and then use `Find File` or `Recent Files` or `Restore Session` to test this. Your keymaps should be working fine.
I found a workaround for this, but am not sure how good a solution this might be. The solution is to reload your `/lua/config/keymaps.lua` file after entering a window. Try the following and see how it works for you and if there is any kind of misbehavior that I might have missed
```lua
vim.api.nvim_create_autocmd("BufWinEnter", {
  group = vim.api.nvim_create_augroup("LazyKeymaps", { clear = true }),
  pattern = "*",
  command = "source /home/jrn23/.config/afzw/lua/config/keymaps.lua",
  desc = "Load Keymaps after plugins",
})
```
Put this in your `/lua/config/autocmds.lua` file.

**Comment by dpetka2001** [comment]
Ok, the solution I wrote in my previous post overwrites the plugins' keymaps completely. So, now you can't scroll through the documentation in the pop up menu of `nvim-cmp`. That's not ideal. I would suggest,to not have any problems, to open Neovim by using only `nvim` without any arguments and use `Find File` or the rest in the `alpha` dashboard menu to open and navigate to files. That way your keymaps in `/lua/config/keymaps.lua` will work and there won't be any conflict with overwriting other plugins' keymaps as well, so I believe you won't lose any functionality.

**Comment by abeldekat** [comment]
LazyVim has defined <c-f> and <c-b> keys as "lazy keys", but Noice is also loaded on VeryLazy.

Perhaps you could unset the keys:
```
  {
    "folke/noice.nvim",
    keys = {
      { "<c-f>", false, mode = { "i", "n", "s" } },
      { "<c-b>", false, mode = { "i", "n", "s" } },
    },
  },
```

**Comment by dpetka2001** [comment]
@abeldekat But without arguments don't they already load before `VeryLazy`? From my testing I observed that the only time they don't get overwritten is when you invoke Neovim without arguments and that's the time they get loaded on `VeryLazy`. So I guess the OP would want his keymaps always loaded on `VeryLazy` even when he invokes Neovim with arguments. Or am I misunderstanding something?

**Comment by abeldekat** [comment]
@dpetka2001, yes, you are right.

Maybe, the keys can be overridden as part of noice:

```lua
return {
  {
    "folke/noice.nvim",
    keys = function()
      local function scroll(movement, key_no_noice_lsp)
        return function()
          if not require("noice.lsp").scroll(movement) then
            return key_no_noice_lsp
          end
        end
      end

      return {
        -- stylua: ignore start
        { "<c-f>", scroll(4, "<c-f>"), silent = true, expr = true, desc = "Scroll forward", mode = { "n", "s" }, },
        { "<c-b>", scroll(-4, "<c-b>"), silent = true, expr = true, desc = "Scroll backward", mode = { "n", "s" }, },

        -- only for insert:
        { "<c-f>", scroll(4, "<RIGHT>"), silent = true, expr = true, desc = "Scroll forward insert", mode = { "i" } },
        { "<c-b>", scroll(-4, "<LEFT>"), silent = true, expr = true, desc = "Scroll backward insert", mode = { "i" } },
        -- stylua: ignore end
      }
    end,
  },
}

```

**Comment by dpetka2001** [comment]
If you tested and it works, I'll put my faith in you :) Let's wait for the OP to tell us if that works for him.

**Comment by afzw** [comment]
First I'd like to thank you guys again for your kind help~~

I have tried all the methods you have provided, and here is the conclusion:

- make an autocmd suggested By @dpetka2001 .

  It really overrided the lazyvim default keymaps, but I can not use the nvim-cmp scroll func.

- nvim without args suggested By @dpetka2001 .

  This way is almost perfect except that I can't put args to fastopen.

- customize the plugin config suggested By @abeldekat

It perform really well, It is really what I need !

------------------ ------------------ boundary ------------------ ------------------

Based on the third way above, now I'd like to use <C-J><C-K> to scroll nvim-cmp popup menu in `insert mode`. And I found that <C-K>(i) was used for "Signature Help popup", I want to rebind it to <C-M>( or sth else ) to use <C-K> for scroll down. Could you please tell me how to do it?

**Comment by dpetka2001** [comment]
```lua
return {
  {
    "hrsh7th/nvim-cmp",
    opts = function(_, opts)
      local cmp = require("cmp")
      opts.mapping = vim.tbl_deep_extend("force", opts.mapping, {
        ["<C-k>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Insert }),
      })
    end,
  },
}
```
It scrolls down to the next item in `nvim-cmp` pop up autocompletion menu. If you want it to do something else, just use the appropriate `nvim-cmp` mapping command. For reference, you can see what Lazyvim uses [here](https://www.lazyvim.org/plugins/coding#nvim-cmp) or you should skim through the `nvim-cmp` docs in case there are other mapping commands that Lazyvim doesn't make use of.

**Comment by dpetka2001** [comment]
@folke I know you closed this as not planned, but I would be interested in your opinion on the more correct approach for how to solve this. I believe @abeldekat 's solution is far better than what I suggested, as mine had several liabilities in the way. I'm just wondering if there's some other way that we couldn't think of. Not implying this is a Lazyvim problem, just a solution for the user experience.

**Comment by folke** [comment]
the correct way to disable the mappings is with:
```
  {
    "folke/noice.nvim",
    keys = {
      { "<c-f>", false, mode = { "i", "n", "s" } },
      { "<c-b>", false, mode = { "i", "n", "s" } },
    },
  },
```

See also here https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps

**Comment by dpetka2001** [comment]
Yes, I know that. Perhaps I wasn't clear on my part. The problem we tried to solve here, is that `/lua/config/keymaps.lua` only get loaded on `VeryLazy` when nvim invoked without arguments. In that case, the keymaps in this thread work just fine, maybe because `keymaps` are the last to load in the `VeryLazy` chain of events in `init.setup()` function, so they don't get overwritten by plugins' keys handlers which are also loaded on `VeryLazy` I would guess? The problem existed only when nvim was invoked with arguments. So the more correct way, according to your opinion, would be to just disable the plugin's keys from the start and not try to change the way the keys handlers behave? Maybe I'm overthinking things due to my lack of knowledge and I apologize for that.

**Comment by folke** [comment]
Should be fixed now. Keymaps are now always lazy loaded

**Comment by dpetka2001** [comment]
Thank you for the swift response and solution @folke. I hope I wasn't perceived as obnoxious from your part. I only wanted to get a better understanding of how things work, regardless what the optimal solution might be. That's why I asked about your opinion in the first place, as I lack knowledge to make such judgements myself and I can only offer some observations I've made (which might or not be correct).

**Comment by folke** [comment]
not at all! and thank you for all the help you offer in this community!

**Comment by davepgreene** [comment]
I know this particular issue is about plugin keymaps but I'm seeing the same issue with remapping the default keys for opening the terminal as well.

My code (`config/keymaps.lua`):
```lua
local function map(mode, lhs, rhs, opts)
  local options = { noremap = true, silent = true }
  if opts then
    options = vim.tbl_extend("force", options, opts)
  end
  vim.keymap.set(mode, lhs, rhs, options)
end

-- Remap terminal toggle
local lazyterm = function()
  Util.float_term(nil, { cwd = Util.get_root() })
end

map("n", "<c-`>", lazyterm, {
  desc = "Terminal (root dir)",
})
map("t", "<c-`>", "<cmd>close<cr>", {
  desc = "Hide Terminal",
})

-- Comment
map("n", "<c-/>", "gcc", {
  desc = "Comment",
})
```

I also set keys in `plugin/comment.lua`:
```lua
return {
  "echasnovski/mini.comment",
  dependencies = {
    "JoosepAlviste/nvim-ts-context-commentstring",
  },
  keys = {
    "gcc",
    "gc",
    { "<c-/>", "gcc", mode = "n", remap = true },
    { "<c-/>", "gc", mode = "v", remap = true },
  },
  opts = function(_, opts)
    return vim.tbl_deep_extend("force", opts, {
      options = {
        ignore_blank_line = true,
        custom_commentstring = function()
          return require("ts_context_commentstring.internal").calculate_commentstring() or vim.bo.commentstring
        end,
      },
    })
  end,
}
```
However, no matter what pressing `ctrl+/` opens the terminal.

I'm more than happy to open a separate issue for this but it seems pretty closely related to how keymaps are loaded.

**Comment by folke** [comment]
That's something totally different. Your terminal probably uses `C-_` to map `C-/`, so change your mappings

**Comment by road2coder** [comment]
@davepgreene Did you resolve the problem?

**Comment by dpetka2001** [comment]
@road2coder What do you want to achieve?

**Comment by road2coder** [comment]
> @road2coder What do you want to achieve?

remap "<c-/>" on windows.(it works well on archlinux)

**Comment by dpetka2001** [comment]
Enter `insert` mode and press `<C-v><C-/>`. What characters are printed out?

**Comment by road2coder** [comment]
> Enter `insert` mode and press `<C-v><C-/>`. What characters are printed out?

`_`

**Comment by road2coder** [comment]
> Enter `insert` mode and press `<C-v><C-/>`. What characters are printed out?

Map it to `^_`  works. But why?
